akka-http-core/src/main/scala/akka/http/impl/engine/rendering/DateHeaderRendering.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.impl.engine.rendering
+
+import akka.actor.{ ClassicActorSystemProvider, Scheduler }
+import akka.annotation.InternalApi
+import akka.http.impl.util.Rendering.CrLf
+import akka.http.impl.util.{ ByteArrayRendering, StringRendering }
+import akka.http.scaladsl.model.{ DateTime, headers }
+
+import java.util.concurrent.atomic.AtomicReference
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContext
+
+/** INTERNAL API */
+@InternalApi private[http] trait DateHeaderRendering {
+  def renderHeaderPair(): (String, String)
+  def renderHeaderBytes(): Array[Byte]
+  def renderHeaderValue(): String
+}
+
+/** INTERNAL API */
+@InternalApi private[http] object DateHeaderRendering {
+  def apply(now: () => Long = () => System.currentTimeMillis())(implicit system: ClassicActorSystemProvider): DateHeaderRendering =
+    apply(system.classicSystem.scheduler, now)(system.classicSystem.dispatcher)
+
+  def apply(scheduler: Scheduler, now: () => Long)(implicit ec: ExecutionContext): DateHeaderRendering = {
+    def renderValue(): String =
+      DateTime(now()).renderRfc1123DateTimeString(new StringRendering).get
+
+    sealed trait DateState
+    /** Date has not been used for a while */
+    case object Idle extends DateState
+    case class AutoUpdated(value: String) extends DateState {
+      var wasUsed: Boolean = false
+      private[this] var _pair: (String, String) = _
+      def headerPair: (String, String) = {
+        if (_pair eq null) _pair = "date" -> value
+        _pair
+      }
+      private[this] var _bytes: Array[Byte] = _
+      def headerBytes: Array[Byte] = {
+        if (_bytes eq null)
+          _bytes = (new ByteArrayRendering(48) ~~ headers.Date ~~ value ~~ CrLf).get
```

Comment:
I wonder if this is really safe. This might be initialized by multiple threads at the same time. `ByteArrayRendering.get` uses `java.util.Arrays.copyOfRange` to fill the result array which is not an atomic operation that establishes a happens-before relationship (in contrast, to creating the tuple above which actually should be safe IIUC).

So, we should either prepopulate the value or make it `@volatile`.

Reply:
Isn't prepopulating fine if we anyway toggle to idle when there is no load? Creating the full header once per second worst case can't be very expensive?

Reply:
Yep, I thought it might make sense because because we often only use one of `headerBytes` or `headerPair`, but shouldn't make a difference whether to render it once per second or not, indeed.